vpr/src/base/vpr_context.h

```diff
@@ -107,6 +107,105 @@ struct hash<std::tuple<int, int, short>> {
 };
 } // namespace std
 
+// MetadataStorage is a two phase data structure.  In the first phase,
+// metadata is added cheaply by simply pushing onto a vector.  This is
+// unsuitable for lookup, but is fast, and because strings are interned, not
+// too expensive memory wise.
+//
+// The second phase occurs after all data has been inserted.  When/if a lookup
+// is needed, the vector is sorted and the final metadata lookup is generated.
+// In the event that the lookup is never needed, this saves the time spent
+// building the lookup, and reduces the number of outstanding memory
+// allocations dramatically.
+template<typename LookupKey>
+class MetadataStorage {
```

Comment:
Generally I've tried to keep the lower-level data structure definitions/details/implementations out of vpr_context.h. Since that header is really key to tracing/finding data in VTR I think its very important to keep it as short and straightforward to traverse as possible.

As a result, I'd suggest moving this class somewhere else. Traditionally this has been `vpr_types.h` but that file is too big. Perhaps move it into a 'metadata_storage.h' in either `vpr/src/base` or `vpr/src/util`?

Reply:
Fixed